### PR TITLE
new: adding files for local php code sniffer using vscode. Resolves #3.

### LINKED ISSUE #3: Add Visual Studio Code local coding standards support
This involves:
- installing extensions
- creating a settings.json file for workspace configuration
- creating a launch.json for using the debugger with PHP
- installing the phpcs and phpcbf dependencies via composer
- creating a .gitignore file to ignore and not commit the local vendor directory created by `composer install`
- a simple bash script to do all of this

### PR DESCRIPTION
**Pull request by AlanBCole**
First pr for vscode setup. _It works locally on my computer_...

### CODE REVIEW

**Review by cathysnider — CHANGES_REQUESTED** [review]
it would be better to append 'vendor' to the .gitignore

File: local_code_sniffer/README.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    2. run `cd ~/.vscode/extensions/ikappas.phpcs-1.0.5/`
+    2. run `git apply ikappas-vscode-phpcs-pullr-172-1.patch`
+5. restart Visual Studio Code.
+
+## Lando setup
+1. In your project's .lando.yml file, make sure that, under `sevices.appserver:`, you have the following:
+    - xdebug: true
+      config:
+        php: .vscode/php.ini
+2. run `lando rebuild` to make the changes in .lando.yml take effect
```
Comment (jwfuller) [inline local_code_sniffer/README.md]: I cannot get this part to work properly
My `.lando.yml` file
```
name: express
recipe: drupal7
config:
  webroot: .
  sevices.appserver:
    xdebug: true
    config:
      php: .vscode/php.ini
```

File: local_code_sniffer/README.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    1. run `mv .vscode/ikappas-vscode-phpcs-pullr-172-1.patch ~/.vscode/extensions/ikappas.phpcs-1.0.5/`
+    2. run `cd ~/.vscode/extensions/ikappas.phpcs-1.0.5/`
+    2. run `git apply ikappas-vscode-phpcs-pullr-172-1.patch`
+5. restart Visual Studio Code.
+
+## Lando setup
+1. In your project's .lando.yml file, make sure that, under `sevices.appserver:`, you have the following:
+    - xdebug: true
+      config:
+        php: .vscode/php.ini
```
Comment (jwfuller) [inline local_code_sniffer/README.md]: Use code highlighting

File: local_code_sniffer/vscode-setup.sh [hunk 3]
```diff
@@ -0,0 +1,41 @@
+rootdir=$1
+firstdir=`pwd`
+
+if [[ $rootdir == '' ]]
+then
+    echo ""
+    echo "What is the full path to your project's root directory? >  "
```
Comment (jwfuller) [inline local_code_sniffer/vscode-setup.sh]:
```suggestion
    echo "What is the full path to your project's Drupal root directory? >  "
```

File: local_code_sniffer/vscode-setup.sh [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+echo -e "vendor\n.vscode\ncomposer.json\ncomposer.lock" >> "$rootdir"/.gitignore
+
+echo ""
+echo "Installing Visual Studio Code extensions ..."
+code --install-extension "felixfbecker.php-debug"
+code --install-extension "felixfbecker.php-intellisense"
+code --install-extension "neilbrayfield.php-docblocker"
+code --install-extension "ikappas.phpcs"
+code --install-extension "persoderlind.vscode-phpcbf"
+code --install-extension "dmitrydorofeev.empty-indent"
```
Comment (jwfuller) [inline local_code_sniffer/vscode-setup.sh:20]: This gives a number of warnings, are there better versions of these extensions?
```
Installing Visual Studio Code extensions ...
Installing extensions...
Extension 'felixfbecker.php-debug' is already installed.
Installing extensions...
Extension 'felixfbecker.php-intellisense' is already installed.
Installing extensions...
Installing extension 'neilbrayfield.php-docblocker' v2.1.0...
(node:81371) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
Extension 'neilbrayfield.php-docblocker' v2.1.0 was successfully installed.
Installing extensions...
Installing extension 'ikappas.phpcs' v1.0.5...
(node:81382) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
Extension 'ikappas.phpcs' v1.0.5 was successfully installed.
Installing extensions...
Installing extension 'persoderlind.vscode-phpcbf' v0.0.8...
(node:81393) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
Extension 'persoderlind.vscode-phpcbf' v0.0.8 was successfully installed.
Installing extensions...
Installing extension 'dmitrydorofeev.empty-indent' v0.2.0...
(node:81404) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
Extension 'dmitrydorofeev.empty-indent' v0.2.0 was successfully installed.
```

**Review by jwfuller — CHANGES_REQUESTED** [review]

File: local_code_sniffer/README.md (hunk 1 above)
Comment (AlanBCole) [reply local_code_sniffer/README.md]: Here is a copy of my .lando.yml file (sorry for the length). You can see where the xdebug and config options are set. _Not_ in the root level `config:` but under 
```yaml
services:  
  appserver:
```
- - -

```yaml 
name: fri-339-12-54
recipe: drupal7
config:
  webroot: .
proxy:
  appserver:
    - fri-339-12-54.lndo.site
services:
  appserver:
    build_internal:
      - >-
        curl
        https://github.com/drush-ops/drush/releases/download/8.3.5/drush.phar -L
        -o /tmp/drush.phar && chmod +x /tmp/drush.phar && mv /tmp/drush.phar
        /usr/local/bin/drush && drush --version
    composer: {}
    config:
      php: .vscode/php.ini
    ssl: true
    type: 'php:7.2'
    via: apache
    xdebug: true
    webroot: .
    overrides:
      environment:
        SIMPLETEST_BASE_URL: 'https://appserver'
        SIMPLETEST_DB: 'mysql://drupal7:drupal7@database/drupal7'
  database:
    config: {}
    authentication: mysql_native_password
    type: mysql
    portforward: true
    creds:
      user: drupal7
      password: drupal7
      database: drupal7
tooling:
  composer:
    service: appserver
    cmd: composer --ansi
  db-import <file>:
    service: ':host'
    description: Imports a dump file into a database service
    cmd: /helpers/sql-import.sh
    user: root
    options:
      host:
        description: The database service to use
        default: database
        alias:
          - h
      no-wipe:
        description: Do not destroy the existing database before an import
        boolean: true
  'db-export [file]':
    service: ':host'
    description: Exports database from a database service to a file
    cmd: /helpers/sql-export.sh
    user: root
    options:
      host:
        description: The database service to use
        default: database
        alias:
          - h
      stdout:
        description: Dump database to stdout
  php:
    service: appserver
    cmd: php
  mysql:
    service: ':host'
    description: Drops into a MySQL shell on a database service
    cmd: mysql -uroot
    options:
      host:
        description: The database service to use
        default: database
        alias:
          - h
  drush:
    service: appserver
```

File: local_code_sniffer/vscode-setup.sh (hunk 4 above)
Comment (AlanBCole) [reply local_code_sniffer/vscode-setup.sh:20]: I have looked at another extension that for both phpcs and phpcbf (called PHP Sniffer & Beautifier valeryanm.vscode-phpsab) which has the same deprecation warnings. I will look into others, as well.

Comment (AlanBCole) [reply local_code_sniffer/vscode-setup.sh:20]: @jwfuller This seems like it is coming from an npm dependency used in the extension and others that I have checked.

**Review by cathysnider — COMMENTED** [review]
Possible improvements to Manual Instructions:

Include update to .gitignore:
echo -e "vendor\n.vscode\ncomposer.json\ncomposer.lock" >> "$rootdir"/.gitignore

CHANGE
1. run `cp composer.json $ROOT`
TO
1. run `cp vscode-dependencies.json $ROOT/composer.json`

CHANGE
`./vendor/bin/phpcs --config-set installed_paths <pathToProjectDirectory>/vendor/drupal/coder/coder_sniffer`
TO
`./vendor/bin/phpcs --config-set installed_paths $ROOT/vendor/drupal/coder/coder_sniffer`

**Review by jwfuller — APPROVED** [review]

**Comment by jwfuller** [comment]
Worked for me, thanks @AlanBCole!

**Review by cathysnider — CHANGES_REQUESTED** [review]
In README.md for localcodesniffer folder, (https://github.com/CuBoulder/cu_debug_admin_bundle/blob/feature/716/local_code_sniffer/README.md) the manual instruction "run localcode sniffer = pwd`` 
should be
`localcodesniffer=$PWD`

**Comment by AlanBCole** [comment]
@cathysnider I put the change in the README.md. I wanted it to be `` localcodesniffer=`pwd` ``. I just did not use the character escaping correctly so it was not rendering correctly. I changed it so it should be showing up correctly now.

**Comment by cathysnider** [comment]
@AlanBCole yes, the back ticks work just as well.

**Review by cathysnider — APPROVED** [review]

**Pull request merged**